src/common/src/hash/key_v2.rs

```diff
@@ -377,6 +375,51 @@ impl<S: KeyStorage, N: NullBitmap> HashKey for HashKeyImpl<S, N> {
     }
 }
 
+/// Helper function to extract the scalar type from an array and returns its
+/// `HashKeySer::exact_size`.
+fn exact_size_of_scalar_in_array<'a, A, S>(_: &A) -> Option<usize>
+where
+    A: Array<RefItem<'a> = S>,
+    S: HashKeySer<'a>,
+{
+    S::exact_size()
+}
+
+#[easy_ext::ext]
+impl DataChunk {
+    fn estimate_hash_key_sizes(&self, column_indices: &[usize]) -> Vec<usize> {
+        let mut estimated_column_indices = Vec::new();
+        let mut exact_size = 0;
+
+        for &i in column_indices {
+            dispatch_all_variants!(&*self.columns()[i], ArrayImpl, col, {
+                match exact_size_of_scalar_in_array(col) {
+                    Some(size) => exact_size += size,
+                    None => estimated_column_indices.push(i),
+                }
+            })
+        }
+
+        let mut sizes = vec![exact_size; self.capacity()];
+
+        for i in estimated_column_indices {
+            dispatch_all_variants!(&*self.columns()[i], ArrayImpl, col, {
+                for ((datum, visible), size) in col
+                    .iter()
+                    .zip_eq_fast(self.vis().iter())
+                    .zip_eq_fast(&mut sizes)
+                {
+                    if visible && let Some(scalar) = datum {
+                        *size += HashKeySer::estimated_size(scalar);
+                    }
+                }
+            })
+        }
```

Comment:
Can we just use `estimated_size` instead of splitting it into 2 blocks?
 My understanding is that if there's an exact size, `estimate_size` would use this exact_size to estimate.

Reply:
The results should be the same, but I'm unsure whether the function call on each scalar can be optimized out. By manually initializing `sizes` with `exact_size`, we're sure the overhead of exact sizes is minimized.